### ISSUE TITLE
Code Testing

### ISSUE THREAD

**Issue by jxu** [llllllllll/slider#43]
Unit tests at the very least to ensure updating the code doesn't break everything

**Comment by fnimick** [comment]
But the ~best~ easiest way to test is to push new code and see if anyone complains about it breaking 🤔

**Comment by llllllllll** [comment]
I agree that slider could use some unit tests, and probably integration/regression tests around the PP calculation and replay parsing. I haven't had time to do that, but I would be open to talking to you about setting that up if you are interested.

**Comment by blimmo** [comment]
I've been pretty busy recently but I should have a bunch of free time starting saturday so I can at the very least review code and probably write some too :)

**Comment by jxu** [comment]
I propose to use the default unittest module and test out basic parts of the API. 
For example: test simple empty beatmap, common beatmaps, tricky beatmaps (old maps, lots of hitobjects, even 2B maps?).
Same goes for testing replays. Ideally the API works for all gamemodes.

**Comment by llllllllll** [comment]
Given that we won't have much complicated persistent state to manage across tests, it's probably easier to use [pytest](https://docs.pytest.org/en/latest/). We use a mix of nose and pytest at work, and I have a slight preference for it.

**Comment by llllllllll** [comment]
NOTE: an old map that has a linear type slider that we once failed to handle correctly, instead treating it like bezier: https://osu.ppy.sh/beatmapsets/19789#osu/69405

**Comment by jxu** [comment]
While we're at it why not test this QA gem from Zallius: "Ã¥Â®â€°Ã¥Â¿Æ’Ã¦â€žÅ¸"
https://osu.ppy.sh/beatmapsets/106#osu/232 

Another one: https://osu.ppy.sh/beatmapsets/104#osu/228 

Minor plug: I've collected a table of pretty much every map loved/ranked up to August 2018 https://github.com/jxu/osu-stats/releases 
IIRC some map had a literal tab character in the tags which of course messed up my TSV writing. And it was a modern map too

**Comment by jxu** [comment]
So I am not familiar with pytest or really "good practice" for testing (most testing I write is informal) 

I imagine the tests to be something like this:

```
import slider

def test_beatmap1():
    # really long string
    beatmap = slider.Beatmap.from_path("data/AKINO from bless4 & CHiCO with HoneyWorks - MIIRO vs. Ai no Scenario (monstrata) [Tatoe].osu")
    assert beatmap.circle_size == 4.0
    assert beatmap.artist_unicode == "AKINO from bless4 & CHiCO with HoneyWorks"
    ...
    
```
For now I'm not sure if it needs to be more complicated than this.
